src/Gaffer/BoxIO.cpp

```diff
@@ -0,0 +1,319 @@
+
+using namespace IECore;
+using namespace Gaffer;
+
+//////////////////////////////////////////////////////////////////////////
+// Internal constants
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+InternedString g_inName( "in" );
+InternedString g_inNamePrivate( "__in" );
+InternedString g_outName( "out" );
+InternedString g_outNamePrivate( "__out" );
+InternedString g_sectionName( "noduleLayout:section" );
+
+std::string oppositeSection( const std::string &section )
+{
+	if( section == "left" )
+	{
+		return "right";
+	}
+	else if( section == "right" )
+	{
+		return "left";
+	}
+	else if( section == "bottom" )
+	{
+		return "top";
+	}
+	else
+	{
+		return "bottom";
+	}
+}
+
+void setupNoduleSectionMetadata( Plug *dst, const Plug *src )
+{
+	ConstStringDataPtr sectionData;
+	for( const Plug *metadataPlug = src; metadataPlug; metadataPlug = metadataPlug->parent<Plug>() )
+	{
+		if( ( sectionData = Metadata::value<StringData>( metadataPlug, g_sectionName ) ) )
+		{
+			break;
+		}
+	}
+
+	if( !sectionData )
+	{
+		return;
+	}
+
+	std::string section = sectionData->readable();
+	if( src->direction() != dst->direction() )
+	{
+		section = oppositeSection( section );
+	}
+
+	Metadata::registerValue( dst, g_sectionName, new StringData( section ) );
+
+}
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// BoxIO
+//////////////////////////////////////////////////////////////////////////
+
+IE_CORE_DEFINERUNTIMETYPED( BoxIO );
+
+size_t BoxIO::g_firstPlugIndex = 0;
+
+BoxIO::BoxIO( Plug::Direction direction, const std::string &name )
+	:	Node( name ), m_direction( direction )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	// Must not accept inputs because the name is syncronised with the promoted
+	// plug name and must therefore not be context-varying.
+	addChild( new StringPlug( "name", Plug::In, direction == Plug::In ? "in" : "out", Plug::Default & ~Plug::AcceptsInputs ) );
+
+	// Connect to the signals we need to syncronise the namePlug() value
+	// with the name of the promotedPlug().
+	plugSetSignal().connect( boost::bind( &BoxIO::plugSet, this, ::_1 ) );
+	if( direction == Plug::In )
+	{
+		plugInputChangedSignal().connect( boost::bind( &BoxIO::plugInputChanged, this, ::_1 ) );
+	}
+	else
+	{
+		parentChangedSignal().connect( boost::bind( &BoxIO::parentChanged, this, ::_2 ) );
+	}
+}
+
+BoxIO::~BoxIO()
+{
+}
+
+StringPlug *BoxIO::namePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const StringPlug *BoxIO::namePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+void BoxIO::setup( const Plug *plug )
+{
+	if( plug )
+	{
+		if( inPlugInternal() )
+		{
+			throw IECore::Exception( "Plugs already set up" );
+		}
+		addChild( plug->createCounterpart( inPlugName(), Plug::In ) );
+		addChild( plug->createCounterpart( outPlugName(), Plug::Out ) );
+
+		inPlugInternal()->setFlags( Plug::Dynamic, true );
```

Comment:
Since @danieldresser hasn't entered the fray yet, I'll launch a preemptive strike...

I do actually share some of Daniel's feelings on this, and it's one of the reasons I've avoided getting to this work for so long, and why I spent so long [trying a different avenue](https://github.com/GafferHQ/gaffer/issues/262#issuecomment-40588991) before settling on this. But despite it not being perfect I do believe that this is the right compromise.

My original approach attempted to avoid any additional nodes at all so that the plug promotion API would remain exactly as it was, and there wouldn't even be a need for "automated re-generation" - I regarded this "pure UI sugar" approach as the perfect solution. But the problem is the users really do want these things to be represented as nodes - they want to select them, position them, delete them, copy/paste them, have them auto-layout etc etc. That means either complicating every single part of the code to deal with these fictional non-node entities, or just compromising and using nodes. So I compromised.

But I don't think then walking halfway back and having the nodes exist or not depending on whether or not they've been viewed is desirable at all - that's not "UI sugar", that's an inconsistent model. Here are just a few of the potential problems :

- We would still need to serialise their position, colour, name etc so we could restore them appropriately when the time came. Where will that be stored if we don't serialise them?
- We would want to see them when viewing References, but we don't want to change the contents of References.
- It would be baffling to drop down a BoxIO node explicitly, then discover it actually doesn't exist any more at a later date, only to find it mysteriously appearing again later.

But maybe you had a different proposal Danie, and I'm barking up the wrong tree?

P.S. This "UI sugar is implemented as persistent nodes" compromise is actually totally consistent with the Dots and Backdrops we already have, so it's nothing new.